Tools/cameras_gimbals/siyi-download/siyi-download.py

```diff
@@ -0,0 +1,142 @@
+            # check response includes 'data'
+            if ('data' not in dir_dict.keys()):
+                exit(prefix_str + "could not get list of directories, no 'data' in response")
+            dir_dict_data = dir_dict['data']
+
+            # check response includes 'directories'
+            if ('directories' not in dir_dict_data.keys()):
+                exit(prefix_str + "could not get list of directories, no 'directories' in response")
+            dir_dict_data_directories = dir_dict_data['directories']
+
+            # create list of directories from 'path' values
+            dir_list = []
+            for dir in dir_dict_data_directories:
+                if ('path' in dir.keys()):
+                    dir_list.append(dir['path'])
+            print(prefix_str + f"{len(dir_list)} directories")
+
+            # get list of files in each directory
+            for dir_path in dir_list:
+                filenames_url = get_filelist_url(ip_address, media_type, dir_path)
+                with urlopen(filenames_url) as get_filenames_url:
+                    filename_dict = json.load(get_filenames_url)
+
+                    # check that the request succeeded
+                    if (not filename_dict['success']):
+                        exit(prefix_str + "failed to get list of files")
+
+                    # check response includes 'data'
+                    if ('data' not in filename_dict.keys()):
+                        exit(prefix_str + "could not get list of files, no 'data' in response")
+                    filename_dict_data = filename_dict['data']
+
+                    # check response includes 'list'
+                    if ('list' not in filename_dict_data.keys()):
+                        exit(prefix_str + "could not get list of files, no 'list' in response")
+                    filename_dict_data_list = filename_dict_data['list']
+                    print(prefix_str + f"{len(filename_dict_data_list)} files")
+
+                    # download each image
+                    for fileinfo in filename_dict_data_list:
+                        if ('name' not in fileinfo.keys() or 'url' not in fileinfo.keys()):
+                            exit(prefix_str + "could not get list of files, no 'name' or 'url' in response")
+                        filename = fileinfo['name']
+                        file_url = fileinfo['url']
+
+                        # correct incorrect ip address in returned url
+                        file_url_fixed = file_url.replace(ip_address_default, ip_address)
+
+                        # download file
+                        print(prefix_str + f"downloading {filename} from {file_url_fixed}")
+                        dest_filename = os.path.join(dest_dir, filename)
+                        try:
+                            urlretrieve(file_url_fixed, dest_filename)
+                        except Exception as e:
+                            print(prefix_str + f"failed to download {filename}: {e}")
```

Comment:
```suggestion
                            print(prefix_str + f"failed to download {filename}: {e}", file=sys.stderr)
```

Reply:
thanks.  I think there is a small typo in the suggestion because it doesn't seem to change the error output